packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/flat_dast/untagged_flat_dast.rs

```diff
@@ -0,0 +1,383 @@
+    pub idx: Index,
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub struct FlatFunctionMacro {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent: Option<Index>,
+    pub path: Vec<PathPart>,
+    pub input: Option<Vec<Vec<UntaggedContent>>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub position: Option<Position>,
+    pub idx: Index,
+}
+
+/// Objects that can be store in the main `nodes` array of a `FlatRoot`.
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "type")]
+pub enum FlatNode {
+    Element(FlatElement),
+    Error(FlatError),
+    FunctionMacro(FlatFunctionMacro),
+    Macro(FlatMacro),
+}
+
+impl Default for FlatNode {
+    fn default() -> Self {
+        FlatNode::Error(FlatError {
+            parent: None,
+            message: "DEFAULT NODE".to_string(),
+            position: None,
+            idx: 0,
+        })
+    }
+}
+
+impl FlatNode {
+    /// Get the index of the node.
+    pub fn idx(&self) -> Index {
+        match self {
+            FlatNode::Element(e) => e.idx,
+            FlatNode::Error(e) => e.idx,
+            FlatNode::FunctionMacro(e) => e.idx,
+            FlatNode::Macro(e) => e.idx,
+        }
+    }
+    /// Get the index of the parent of the node.
+    pub fn parent(&self) -> Option<Index> {
+        match self {
+            FlatNode::Element(e) => e.parent,
+            FlatNode::Error(e) => e.parent,
+            FlatNode::FunctionMacro(e) => e.parent,
+            FlatNode::Macro(e) => e.parent,
+        }
+    }
+}
+
+/// Untagged version of a Flattened DAST. All elements/errors/macros/function macros are stored in
+/// the `nodes` vec. All children of all elements/attributes are vectors of text or references to
+/// positions in the `nodes` vec.
+///
+/// These references are untagged, so the type of each node may be mutated and the reference remains valid.
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "type")]
+pub struct FlatRoot {
+    pub children: Vec<UntaggedContent>,
+    pub nodes: Vec<FlatNode>,
+}
+
+impl FlatRoot {
+    pub fn new() -> Self {
+        Self {
+            children: Vec::new(),
+            nodes: Vec::new(),
+        }
+    }
+
+    /// Create a new `FlatRoot` from a `DastRoot`.
+    pub fn from_dast(dast: &DastRoot) -> Self {
+        let mut flat = Self::new();
+        flat.merge_dast_root(dast);
+        flat
+    }
+
+    /// Iterate over the parent elements of a node.
+    /// If for some reason the node has a non-element parent, the iterator will panic.
+    pub fn parent_iter(&self, start_idx: Index) -> ParentIterator {
+        let start = &self.nodes[start_idx];
+        ParentIterator::new(start, self)
+    }
+
+    /// Merge the content of a `DastRoot` into `FlatRoot`.
+    /// This function recursively adds all children etc. of the `DastRoot`.
+    pub fn merge_dast_root(&mut self, dast: &DastRoot) {
+        for child in dast.children.iter() {
+            self.merge_content(child, None);
```

Comment:
If you collect these results into the `self.children` array here, you wouldn't have to special case `parent.is_none()` in `merge_content()`.

Reply:
That's true, but if I set `flat_root.children` here, I wouldn't be able to do a `merge_content` with parent set to `None` and have them show up in Root. I don't know if I ever need that functionality though...(maybe it's better to not have it since passing an otherwise invalid parent does nothing?)